### PR TITLE
[NRL-796] Require X-Request-Id and NHSD-Correlation-Id in all requests for lambda log correlation

### PR DESCRIPTION
**Pull request by mattdean3-nhs**

### CODE REVIEW

File: layer/nrlf/core/constants.py [hunk 1]
```diff
@@ -34,6 +34,12 @@ class Source(Enum):
 PERMISSION_ALLOW_ALL_POINTER_TYPES = "allow-all-pointer-types"
 
 
+NHSD_REQUEST_ID_HEADER = "NHSD-Request-ID"
+NHSD_CORRELATION_ID_HEADER = "NHSD-Correlation-Id"
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/core/constants.py:38]: ID case

File: layer/nrlf/core/decorators.py [hunk 2]
```diff
@@ -98,6 +172,30 @@ def filter_kwargs(handler_func: RequestHandler, kwargs: Dict[str, Any]):
     return function_kwargs
 
 
+def verify_request_ids(event: APIGatewayProxyEvent):
+    caller_request_id = event.get_header_value("x-request-id")
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/core/decorators.py]: Use constant

File: layer/nrlf/core/decorators.py [hunk 3]
```diff
@@ -98,6 +172,30 @@ def filter_kwargs(handler_func: RequestHandler, kwargs: Dict[str, Any]):
...
+        logger.log(LogReference.HANDLER014, headers=event.headers)
+        raise OperationOutcomeError(
+            status_code="400",
+            severity="error",
+            code="invalid",
+            details=SpineErrorConcept.from_code("MISSING_OR_INVALID_HEADER"),
+            diagnostics="The X-Request-Id header is missing or invalid",
+        )
+
+    caller_correlation_id = event.get_header_value("nhsd-correlation-id")
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/core/decorators.py]: Use constant for id

File: layer/nrlf/core/decorators.py [hunk 4]
```diff
@@ -194,10 +295,11 @@ def wrapper(*args, **kwargs):
             return response.dict()
 
         decorators = [
-            wraps(func),
-            logger.inject_lambda_context,
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/core/decorators.py:198]: Put this back in or investigate

File: layer/nrlf/tests/events.py [hunk 5]
```diff
@@ -22,11 +29,20 @@ def create_headers(
         "nhsd-client-rp-details": json.dumps(
             {"developer.app.name": app_name, "developer.app.id": app_id}
         ),
+        "X-Request-Id": "test_request_id",
+        "NHSD-Correlation-Id": "test_correlation_id",
+        **addional_headers,
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/tests/events.py]: Typo

File: layer/nrlf/tests/events.py [hunk 6]
```diff
@@ -3,12 +3,19 @@
...
+    }
+
+
 def create_headers(
     ods_code: str = "Y05868",
     nrl_permissions: Optional[List[str]] = [],
     app_name: Optional[str] = "TestApp",
     app_id: Optional[str] = "12345",
     nrl_app_id: Optional[str] = "Y05868-TestApp-12345678",
+    addional_headers: Optional[Dict[str, str]] = {},
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/tests/events.py]: Typo

File: layer/nrlf/core/tests/test_decorators.py [hunk 7]
```diff
@@ -108,6 +119,226 @@ def decorated_function():
...
+
+    test_event = create_test_api_gateway_event()
+    event = APIGatewayProxyEvent(test_event)
+
+    decorated_function(event)
+
+    mock_logger.set_correlation_id.assert_called_once_with("test_correlation_id")
+
+
+def test_logger_initialiser_no_correletion_id(mocker: MockerFixture):
```
Comment (axelkrastek1-nhs) [inline layer/nrlf/core/tests/test_decorators.py]: typo

**Review by axelkrastek1-nhs — COMMENTED** [review]
Add a check of headers to integrations test happy path

**Review by katebobyn-nhs — APPROVED** [review]

**Pull request merged**